parlai/agents/transformer/polyencoder.py

```diff
@@ -0,0 +1,267 @@
+        agent.add_argument('--polyencoder-type', type=str, default='codes',
+                           choices=['codes', 'n_first'],
+                           help='Type of polyencoder, either we compute'
+                                'vectors using codes + attention, or we '
+                                'simply take the first N vectors.')
+        agent.add_argument('--poly-n-codes', type=int, default=64,
+                           help='number of vectors used to represent the context'
+                                'in the case of n_first, those are the number'
+                                'of vectors that are considered.')
+        agent.add_argument('--poly-attention-type', type=str, default='basic',
+                           choices=['basic', 'basic_sqrt', 'multihead'],
+                           help='Type of the top aggregation layer of the poly-'
+                                'encoder (where the candidate representation is'
+                                'the key)')
+        agent.add_argument('--poly-attention-num-heads', type=int, default=4,
+                           help='In case poly-attention-type is multihead, '
+                                'specify the number of heads')
+
+        # Those arguments are here in case where polyencoder type is 'code'
+        agent.add_argument('--codes-attention-type', type=str, default='basic',
+                           choices=['basic', 'basic_sqrt', 'multihead'],
+                           help='Type ')
+        agent.add_argument('--codes-attention-num-heads', type=int, default=4,
+                           help='In case codes-attention-type is multihead, '
+                                'specify the number of heads')
+        return agent
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.rank_loss = torch.nn.CrossEntropyLoss(reduce=True, size_average=True)
+        if self.use_cuda:
+            self.rank_loss.cuda()
+        self.data_parallel = opt.get('data_parallel') and self.use_cuda
+        if self.data_parallel:
+            from parlai.core.distributed_utils import is_distributed
+            if is_distributed():
+                raise ValueError(
+                    'Cannot combine --data-parallel and distributed mode'
+                )
+            self.model = torch.nn.DataParallel(self.model)
+        self.START_IDX = self.END_IDX
+
+
+    def build_model(self, states=None):
+        self.model = PolyEncoderModule(self.opt, self.dict, self.NULL_IDX)
+        return self.model
+
+    def vectorize(self, *args, **kwargs):
+        kwargs['add_start'] = True
+        kwargs['add_end'] = True
+        obs = super().vectorize(*args, **kwargs)
+        return obs
+
+    def _set_text_vec(self, *args, **kwargs):
+        """ Add start and end tokens for set_text_vec()
+        """
+        obs = super()._set_text_vec(*args, **kwargs)
+        if 'text_vec' in obs:
```

Comment:
you can also call `_add_start_end_tokens` from torch_agent

Reply:
This method seems to expect a python list though. Maybe i should rewrite it so that it also handle tensors?

Reply:
ooo good call... yes i think that might be helfpul. thoughts @stephenroller ?

Reply:
I updated _add_start_end_tokens to handle tensors